core/src/main/java/com/backblaze/b2/json/B2JsonBoundedByteArrayOutputStream.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2021, Backblaze Inc. All Rights Reserved.
+ * License https://www.backblaze.com/using_b2_code.html
+ */
+package com.backblaze.b2.json;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.Arrays;
+
+/**
+ * A B2Json implementation of OutputStream that
+ * - stores stream content in a byte array
+ * - expands capacity when needed (typically doubles)
+ * - has a max capacity threshold
+ * - throws IOException if the threshold is reached
+ *
+ * THREAD-SAFE
+ */
+public class B2JsonBoundedByteArrayOutputStream extends OutputStream {
+    private static final int INITIAL_CAPACITY = 64;
+
+    // byte array to hold output content
+    private byte[] output;
+
+    // the current number of bytes written so far
+    private int size;
+
+    // maximum capacity the output array is allowed to grow to
+    private static final int MAX_CAPACITY = Integer.MAX_VALUE - 8;
+
+    public B2JsonBoundedByteArrayOutputStream() {
```

Comment:
As mentioned below, I think the constructor should take the maximum buffer size, rather than using a constant.

That will make it easier to read calling code and know what it's doing.